### PR TITLE
Convert aprox21 to C++

### PR DESCRIPTION
**Pull request by maxpkatz**

### CODE REVIEW

**Comment by maxpkatz** [comment]
Note: #322 and #323 are necessary to be merged for this C++ implementation to exactly match the Fortran implementation.

**Comment by shardi2** [comment]
Mass_Fractions_plots_aprox21

**Comment by shardi2** [comment]
Mass_Fractions_plots_aprox21

**Comment by shardi2** [comment]
Mass_Fractions_plots_aprox21

**Comment by zingale** [comment]
test_react comparison:
```
$ fcompare.gnu.ex react_aprox21_test_react.VODE react_aprox21_test_react.VODE.cxx

            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                                          0                         0
 temperature                                      0                         0
 Xnew_hydrogen-1                    4.596057776e-09           1.094511334e-07
 Xnew_helium-3                      1.569385087e-09           4.708155486e-08
 Xnew_helium-4                      1.392332782e-08           1.971051007e-08
 Xnew_carbon-12                     3.071432668e-08           4.387760954e-08
 Xnew_nitrogen-14                    3.58257364e-08           1.423074706e-07
 Xnew_oxygen-16                      4.90626384e-09           6.974884598e-09
 Xnew_neon-20                       8.437301899e-09           3.981746643e-08
 Xnew_magnesium-24                   8.92546375e-09            6.02587633e-08
 Xnew_silicon-28                    1.276977316e-08           2.262364773e-08
 Xnew_sulfur-32                     1.347081204e-08           7.445618157e-08
 Xnew_argon-36                      1.174121243e-08            1.10031356e-07
 Xnew_calcium-40                    1.474224734e-08           5.823411667e-08
 Xnew_titanium-44                   1.055154167e-09           3.163041348e-08
 Xnew_chromium-48                   6.162228228e-10            1.79734175e-08
 Xnew_chromium-56                   3.564856743e-14           1.069457023e-12
 Xnew_iron-52                       1.740342911e-09            3.41773696e-08
 Xnew_iron-54                       1.313251058e-08           3.821804064e-08
 Xnew_iron-56                       7.364877784e-09           3.547273151e-08
 Xnew_nickel-56                     1.912868091e-09           5.720173951e-08
 Xnew_neutron                        1.89341369e-09           5.680241069e-08
 Xnew_proton                        1.894106469e-09           5.505723386e-08
 Xold_hydrogen-1                                  0                         0
 Xold_helium-3                                    0                         0
 Xold_helium-4                                    0                         0
 Xold_carbon-12                                   0                         0
 Xold_nitrogen-14                                 0                         0
 Xold_oxygen-16                                   0                         0
 Xold_neon-20                                     0                         0
 Xold_magnesium-24                                0                         0
 Xold_silicon-28                                  0                         0
 Xold_sulfur-32                                   0                         0
 Xold_argon-36                                    0                         0
 Xold_calcium-40                                  0                         0
 Xold_titanium-44                                 0                         0
 Xold_chromium-48                                 0                         0
 Xold_chromium-56                                 0                         0
 Xold_iron-52                                     0                         0
 Xold_iron-54                                     0                         0
 Xold_iron-56                                     0                         0
 Xold_nickel-56                                   0                         0
 Xold_neutron                                     0                         0
 Xold_proton                                      0                         0
 wdot_hydrogen-1                       0.4596057757           1.522788613e-07
 wdot_helium-3                         0.1569385068           4.708156501e-08
 wdot_helium-4                          1.392332871           3.849214943e-08
 wdot_carbon-12                         3.071432653           4.397100617e-08
 wdot_nitrogen-14                       3.582573642            1.58971197e-07
 wdot_oxygen-16                        0.4906263838           7.108275696e-09
 wdot_neon-20                          0.8437301898           4.231355236e-08
 wdot_magnesium-24                     0.8925463769           6.260698856e-08
 wdot_silicon-28                        1.276977319           2.284853529e-08
 wdot_sulfur-32                         1.347081202           7.681492147e-08
 wdot_argon-36                          1.174121241           1.246308987e-07
 wdot_calcium-40                        1.474224731           6.125889062e-08
 wdot_titanium-44                      0.1055154164           3.167112317e-08
 wdot_chromium-48                     0.06162228193           1.849411081e-08
 wdot_chromium-56                   3.564856743e-06              0.9577740492
 wdot_iron-52                          0.1740342914           3.836281602e-08
 wdot_iron-54                           1.313251056           3.884609242e-08
 wdot_iron-56                           0.736487777           3.644801398e-08
 wdot_nickel-56                        0.1912868088            5.73925009e-08
 wdot_neutron                          0.1893413689           5.680327755e-08
 wdot_proton                           0.1894106469            6.19355351e-08
 rho_Hnuc                           1.076873937e+26           1.192378853e-08
```

**Comment by zingale** [comment]
Note: that comparison has the most recent `development` merged into this PR, so the integrator is slightly different for C++ now.

**Comment by zingale** [comment]
The Cr56 difference looks a bit large

**Comment by zingale** [comment]
if I don't merge in development, this is the difference:
```
fcompare.gnu.ex react_aprox21_test_react.VODE react_aprox21_test_react.VODE.cxx

            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                                          0                         0
 temperature                                      0                         0
 Xnew_hydrogen-1                    3.404883772e-09           8.108435671e-08
 Xnew_helium-3                        1.1597046e-09           3.479113967e-08
 Xnew_helium-4                      1.073724071e-08           1.520013705e-08
 Xnew_carbon-12                     2.274735685e-08           3.249622406e-08
 Xnew_nitrogen-14                   2.653372449e-08           1.053976163e-07
 Xnew_oxygen-16                     3.522067638e-09           5.007071801e-09
 Xnew_neon-20                       5.824615014e-09           2.748762752e-08
 Xnew_magnesium-24                  7.274333646e-09           4.911143688e-08
 Xnew_silicon-28                    1.115115972e-08           1.975602119e-08
 Xnew_sulfur-32                     1.097623897e-08           6.066812006e-08
 Xnew_argon-36                      9.612344296e-09           9.008092516e-08
 Xnew_calcium-40                    8.020238966e-09           3.168116236e-08
 Xnew_titanium-44                   5.506857297e-10           1.650793587e-08
 Xnew_chromium-48                   4.674841486e-10           1.363514539e-08
 Xnew_chromium-56                   2.465389004e-14           7.396167012e-13
 Xnew_iron-52                       6.625547186e-10           1.301144582e-08
 Xnew_iron-54                       1.250941617e-08           3.640472039e-08
 Xnew_iron-56                       2.985134417e-09           1.437781791e-08
 Xnew_nickel-56                     6.525422459e-10           1.951339548e-08
 Xnew_neutron                       8.979699068e-10            2.69390972e-08
 Xnew_proton                        8.982930962e-10           2.611127404e-08
 Xold_hydrogen-1                                  0                         0
 Xold_helium-3                                    0                         0
 Xold_helium-4                                    0                         0
 Xold_carbon-12                                   0                         0
 Xold_nitrogen-14                                 0                         0
 Xold_oxygen-16                                   0                         0
 Xold_neon-20                                     0                         0
 Xold_magnesium-24                                0                         0
 Xold_silicon-28                                  0                         0
 Xold_sulfur-32                                   0                         0
 Xold_argon-36                                    0                         0
 Xold_calcium-40                                  0                         0
 Xold_titanium-44                                 0                         0
 Xold_chromium-48                                 0                         0
 Xold_chromium-56                                 0                         0
 Xold_iron-52                                     0                         0
 Xold_iron-54                                     0                         0
 Xold_iron-56                                     0                         0
 Xold_nickel-56                                   0                         0
 Xold_neutron                                     0                         0
 Xold_proton                                      0                         0
 wdot_hydrogen-1                       0.3404883753           1.128122944e-07
 wdot_helium-3                         0.1159704581           3.479114703e-08
 wdot_helium-4                          1.073723981           2.968395332e-08
 wdot_carbon-12                         2.274735669           3.256539454e-08
 wdot_nitrogen-14                       2.653372451           1.177393228e-07
 wdot_oxygen-16                        0.3522067789           5.102829708e-09
 wdot_neon-20                          0.5824615015           2.921077785e-08
 wdot_magnesium-24                     0.7274333626           5.102526146e-08
 wdot_silicon-28                        1.115115972           1.995240342e-08
 wdot_sulfur-32                         1.097623899           6.259006023e-08
 wdot_argon-36                         0.9612344312           1.020333395e-07
 wdot_calcium-40                       0.8020238969           3.332673311e-08
 wdot_titanium-44                     0.05506857298           1.652918235e-08
 wdot_chromium-48                     0.04674841475           1.403015817e-08
 wdot_chromium-56                   2.465389004e-06              0.6625023308
 wdot_iron-52                         0.06625547176           1.460486007e-08
 wdot_iron-54                           1.250941616           3.700297319e-08
 wdot_iron-56                          0.2985134404           1.477311965e-08
 wdot_nickel-56                       0.06525422423           1.957847039e-08
 wdot_neutron                         0.08979699062           2.693950832e-08
 wdot_proton                          0.08982930961           2.937335601e-08
 rho_Hnuc                           7.448363978e+25           8.247271463e-09
```

**Comment by zingale** [comment]
The Cr56 stuff might just be insignificant.

**Comment by zingale** [comment]
If I change the probin.aprox21 to this:
```
&extern

  dens_min   = 1.d4
  dens_max   = 1.d8
  temp_min   = 5.d7
  temp_max   = 5.d9

  tmax = 1.d-8

  primary_species_1 = "hydrogen-1"
  primary_species_2 = "helium-4"
  primary_species_3 = "oxygen-16"


  jacobian = 1
  centered_diff_jac = T

  rtol_spec = 1.d-8
  atol_spec = 1.d-8

  call_eos_in_rhs = T
  do_constant_volume_burn = T

/
```

(mainly, add some H to the net), then I see really large differences:
```
$ fcompare.gnu.ex react_aprox21_test_react.VODE react_aprox21_test_react.VODE.cxx

            variable name            absolute error            relative error
                                        (||A - B||)         (||A - B||/||A||)
 ----------------------------------------------------------------------------
 level = 0
 density                                          0                         0
 temperature                                      0                         0
 Xnew_hydrogen-1                        0.002455499            0.003502158478
 Xnew_helium-3                       2.80851241e-05           0.0008425537611
 Xnew_helium-4                         0.2498368642              0.3526964251
 Xnew_carbon-12                     8.162397885e-05            0.002448719366
 Xnew_nitrogen-14                     0.05485594231              0.7599938598
 Xnew_oxygen-16                       0.07041982309              0.1003302853
 Xnew_neon-20                         0.03321760907              0.5744864686
 Xnew_magnesium-24                    0.02535056535              0.4988641176
 Xnew_silicon-28                      0.08935495468              0.1836834602
 Xnew_sulfur-32                       0.05612606444              0.4478097311
 Xnew_argon-36                        0.04831457537              0.5377032351
 Xnew_calcium-40                       0.1016258239              0.5058746824
 Xnew_titanium-44                     0.02260099337              0.6769099793
 Xnew_chromium-48                     0.01823757278              0.4896033778
 Xnew_chromium-56                    2.95651136e-07            8.86953298e-06
 Xnew_iron-52                         0.01944444447              0.3622249428
 Xnew_iron-54                          0.2930105249              0.7673714273
 Xnew_iron-56                         0.07338962187              0.3892984338
 Xnew_nickel-56                       0.03662119799              0.6644274676
 Xnew_neutron                         0.02638847542              0.7916542627
 Xnew_proton                          0.02439488655              0.7091042867
 Xold_hydrogen-1                                  0                         0
 Xold_helium-3                                    0                         0
 Xold_helium-4                                    0                         0
 Xold_carbon-12                                   0                         0
 Xold_nitrogen-14                                 0                         0
 Xold_oxygen-16                                   0                         0
 Xold_neon-20                                     0                         0
 Xold_magnesium-24                                0                         0
 Xold_silicon-28                                  0                         0
 Xold_sulfur-32                                   0                         0
 Xold_argon-36                                    0                         0
 Xold_calcium-40                                  0                         0
 Xold_titanium-44                                 0                         0
 Xold_chromium-48                                 0                         0
 Xold_chromium-56                                 0                         0
 Xold_iron-52                                     0                         0
 Xold_iron-54                                     0                         0
 Xold_iron-56                                     0                         0
 Xold_nickel-56                                   0                         0
 Xold_neutron                                     0                         0
 Xold_proton                                      0                         0
 wdot_hydrogen-1                           245549.9             0.01592430422
 wdot_helium-3                           2808.51241           0.0008425539271
 wdot_helium-4                          24983686.42                0.42736449
 wdot_carbon-12                         8162.397885            0.002448719366
 wdot_nitrogen-14                       5485594.231               1.412134352
 wdot_oxygen-16                         7041982.309              0.1056901636
 wdot_neon-20                           3321760.907               1.007199251
 wdot_magnesium-24                      2535056.535              0.7738242967
 wdot_silicon-28                        8935495.468              0.1879769811
 wdot_sulfur-32                         5612606.444              0.5251367103
 wdot_argon-36                          4831457.537              0.8296542768
 wdot_calcium-40                        10162582.39              0.5823748895
 wdot_titanium-44                       2260099.337              0.6780856386
 wdot_chromium-48                       1823757.278              0.5472005031
 wdot_chromium-56                        29.5651136               1.000000001
 wdot_iron-52                           1944444.447              0.4104204171
 wdot_iron-54                           29301052.49              0.7933427009
 wdot_iron-56                           7338962.187              0.4011192932
 wdot_nickel-56                         3662119.799               0.738906116
 wdot_neutron                           2638847.542              0.7916545796
 wdot_proton                            2439488.655              0.7711093976
 rho_Hnuc                           6.963146083e+33               0.658559294
```

File: networks/aprox21/actual_rhs.H [hunk 1]
```diff
@@ -0,0 +1,3009 @@
...
+{
+    rate_t rr;
+
+    bool deriva;
+
+    Real b1, sneut, dsneutdt, dsneutdd, snuda, snudz;
+
+    Real rho, temp, abar, zbar;
+    Array1D<Real, 1, NumSpec> y;
+    Array1D<Real, 1, NumRates> r1, r2, r3;
```
Comment (harpolea) [inline networks/aprox21/actual_rhs.H:2939]:
```suggestion
```

File: networks/aprox21/actual_rhs.H [hunk 2]
```diff
@@ -0,0 +1,3009 @@
...
+        y(n) = state.xn[n-1] * aion_inv[n-1];
+    }
+
+    // Species Jacobian elements with respect to other species
+
+    for (int i = 1; i <= NumRates; ++i) {
+        r1(i) = rr.rates(1, i);
+        r2(i) = rr.rates(3, i);
+        r3(i) = rr.rates(4, i);
+    }
```
Comment (harpolea) [inline networks/aprox21/actual_rhs.H:2965]:
```suggestion
```

File: networks/aprox21/actual_rhs.H [hunk 3]
```diff
@@ -0,0 +1,3009 @@
...
+        jac.add(net_ienuc, j, -b1);
+    }
+
+    // Evaluate the Jacobian elements with respect to temperature by
+    // calling the RHS using d(ratdum) / dT
+
+    for (int i = 1; i <= NumRates; ++i) {
+        r1(i) = rr.rates(1, i);
+        r2(i) = rr.rates(2, i);
+    }
```
Comment (harpolea) [inline networks/aprox21/actual_rhs.H:2991]:
```suggestion
```

File: networks/aprox21/actual_rhs.H [hunk 4]
```diff
@@ -0,0 +1,3009 @@
...
+    for (int i = 1; i <= NumSpec; ++i) {
+        y(i) = state.xn[i-1] * aion_inv[i-1];
+    }
+
+    // Call the RHS to actually get dydt.
+
+    for (int i = 1; i <= NumRates; ++i) {
+        r1(i) = rr.rates(1, i);
+        r2(i) = rr.rates(1, i);
+    }
```
[the remark this answers is not in the document]
Comment (maxpkatz) [reply networks/aprox21/actual_rhs.H]: Removed

**Comment by maxpkatz** [comment]
>If I change the probin.aprox21 to this:
> ...
> (mainly, add some H to the net), then I see really large differences

I can still reproduce this but strangely these large diffs don't appear in test_rhs with the same composition.

**Review by zingale — APPROVED** [review]

**Pull request merged**